lms/djangoapps/course_api/permissions.py

```diff
@@ -2,9 +2,65 @@
 Course API Authorization functions
 """
 
+from django.contrib.auth.models import User
+from rest_framework import permissions
+
 from student.roles import GlobalStaff
 
 
+class MasqueradingPermission(permissions.BasePermission):
+    """
+    Allow authorized users to masquerade as other users, taking on that
+    user's permissions.
+
+    This class must be listed before any other Permission classes
+    to properly enable masquerading
+
+    To masquerade as a user, add a query parameter to the URL specifying the
+    masqueraded user's username:
+
+        GET https://courses.edx.org/api/courses/v1/courses?masquerade=henry
+
+    To use a parameter other than `masquerade`, set `masquerading_param` as a
+    class attribute on any view that uses `MasqueradingPermission`.
+    """
+
+    default_masquerading_param = 'masquerade'
+
+    def has_permission(self, request, view):
+        """
+        Hook for running masquerade.
+
+        TODO: clarify what is happening
+        """
+        masquerading_param = getattr(view, 'masquerading_param', None) or self.default_masquerading_param
+        masqueraded_username = request.query_params.get(masquerading_param, request.user.username)
+        if request.user.username == masqueraded_username:
+            return True
+        elif self._can_masquerade(request.user):
+            return self._masquerade(request, masqueraded_username)
+        else:
+            return False
+
+    def _can_masquerade(self, user):
+        """
+        Determine whether the user has permission to masquerade as other users
+        """
+        staff = GlobalStaff()
+        return staff.has_user(user)
+
+    def _masquerade(self, request, username):
+        """
+        Enable masquerading.  The effective user is now the masqueraded one.
+        """
+        masqueraded_user = User.objects.get(username=username)
+        if masqueraded_user:
+            request.user = masqueraded_user
```

Comment:
Authentication is complete by the time permission checks are run, so there shouldn't be any issues with reassigning request.user at this point in the request-response cycle.